libs/chain-signatures/contract/src/lib.rs

```diff
@@ -193,6 +204,37 @@ impl MpcContract {
         }
         Ok(())
     }
+
+    pub fn vote_code_hash(&mut self, code_hash: CodeHash) -> Result<(), Error> {
+        let code_hash_votes = self
+            .protocol_state
+            .proposed_code_hashes_count_votes(code_hash.clone())?;
+
+        if code_hash_votes + 1 >= self.threshold()?.value() {
+            self.protocol_state.clear_code_hashes_votes()?;
+            self.historical_code_hashes.push(code_hash.clone());
+            self.allowed_code_hashes.push(code_hash.clone());
+            self.upgrade_deadline = Some(env::block_height() + TEE_UPGRADE_PERIOD);
+        } else if let Some(new_state) = self.protocol_state.vote_code_hash(code_hash)? {
+            self.protocol_state = new_state;
+        }
+
+        Ok(())
+    }
+
+    pub fn allowed_code_hashes(&mut self) -> Vec<CodeHash> {
+        if env::block_height() > self.upgrade_deadline.unwrap_or_default() {
+            self.allowed_code_hashes = self
+                .allowed_code_hashes
+                .last()
+                .map_or_else(Vec::new, |last| vec![last.clone()]);
+        }
+        self.allowed_code_hashes.clone()
+    }
+
+    pub fn latest_code_hash(&self) -> Option<CodeHash> {
+        self.allowed_code_hashes.last().cloned()
```

Comment:
I think we should return `CodeHash`. The invariant is that there should always be a latest code hash, i.e, it is not possible for it to be `None`.

Reply:
yes it is safe